OIDC RP-Initiated logout endpoint

mposolda opened the pull request:
Closes #10885

Docs PR: https://github.com/keycloak/keycloak-documentation/pull/1443
NodeJS PR: https://github.com/keycloak/keycloak-nodejs-connect/pull/359

I've checked that with these changes, we should be able to certify OIDC logout with all the profiles (RP-Initiated, Frontchannel, Backchannel, Session management) 

Pipeline runs:
https://keycloak-jenkins.com/job/universal-test-pipeline-server/1333/
https://keycloak-jenkins.com/job/universal-test-pipeline-adapters/926/
I've checked pipeline errors are either unrelated or already fixed in the latest branch. With the exception of the node.js 3 test failures in the adapter testsuite, which I  need to fix in the separate PR to the node.js repository.

douglaspalmer committed:
OIDC RP-Initiated logout endpoint
Closes #10885

Co-Authored-By: Marek Posolda <mposolda@gmail.com>

stianst commented:
Started looking at the code, and looks good so far, also tried this out a bit manually.

One thing that surprised me was 400 on missing id_token_hint. I was actually expecting this to present the logout confirmation page (without redirect or back to app link). The spec states id_token_hint is recommended, but not required.

stianst commented on services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java:
```diff
@@ -154,7 +153,18 @@
+
+    // ** Logout related notes **/
+    // Flag in the logout session to specify if we use "system" client or real client
+    public static final String LOGOUT_WITH_SYSTEM_CLIENT = "LOGOUT_WITH_SYSTEM_CLIENT";
```
Not quite following the "system" client scenario. Could you elaborate on this one?

mposolda replied:
I was thinking about removing this constant during the PR cleanup, will probably do that.

The point of the "system" client is, that AuthenticationSessionModel always need to have client, which it is linked too. In normal OIDC/SAML login scenarios, authentication session always has client. But in some rare cases (ActionTokens, or now Logout) there is not dedicated client of the flow and hence the "system" client is used as a placeholder. The system client is currently "account" client.

From the long term perspective, it may be better to remove the need of "system client" or use dedicated client just for that instead of the "account" client.

In the case of the system client, the link "Back to application" should not be shown on the pages as it is not correct to show "Back to application" to point to account management in case that "account" was used only as a placeholder system client. Hence I've added the session note to differentiate between the cases when "account" was used as a "system" client or as real client during logout from account console. However session note is not needed as logout from account console is not a problem because:

- New account console uses "account-console" client
- Old account console uses "account" client, but user is immediately logged out when click to "Logout"

So I can remove the note and instead directly check if the client is the system client (account) to decide if the link "Back to application" should be shown.

mposolda replied:
Updated to remove this session note and instead directly check if client is "system client" to doublecheck if the link "Back to application" should be shown or not.

stianst commented:
Looks good to me, just need one thing explained to be, see previous comment.

We should already now schedule the removal of the legacy support for `redirect_uri` and I suggest we do that in Keycloak 21. See https://github.com/keycloak/keycloak/issues/10983 and https://www.keycloak.org/2022/03/releases.html.

Could you update https://github.com/keycloak/keycloak/issues/10983 with some notes on what should be removed, so we have some hints in the future what work is needed?

Also, would be good to update the changes in docs PR to make it explicitly that this will be removed in the future.

mposolda commented on services/src/main/java/org/keycloak/protocol/oidc/OIDCProviderConfig.java:
```diff
@@ -0,0 +1,37 @@
+    private final Config.Scope config;
+
+    public OIDCProviderConfig(Config.Scope config) {
+        this.config = config;
```
Yes, true. I've used the current way just because of the tests, however it is probably better to update the test to not require resolving this every time in production. I will update this.

mposolda replied:
Updated

mposolda committed:
Review feedback

mposolda commented:
@stianst I've updated PR per your feedback. I've used separate commit, so you can easily check just the changes from the original version, but I plan to squash the commits before merging.

Updated follow-up issue https://github.com/keycloak/keycloak/issues/10983 with the steps for removing backwards compatibility.

akikoskinen commented:
> Never mind, checked the spec again and `id_token_hint` is required when `post_logout_redirect_uri` is included.

@stianst @mposolda Is this the spec https://openid.net/specs/openid-connect-rpinitiated-1_0.html?

I don't find text in the actual specification that would say that `id_token_hint` would be required. It's just recommended, in various places. There is one mention of it being required, but that's in the [document history](https://openid.net/specs/openid-connect-rpinitiated-1_0.html#History) section. But AFAIK the document history section isn't part of the actual specification.

mposolda commented:
> > Never mind, checked the spec again and `id_token_hint` is required when `post_logout_redirect_uri` is included.
> 
> @stianst @mposolda Is this the spec https://openid.net/specs/openid-connect-rpinitiated-1_0.html?
> 
> I don't find text in the actual specification that would say that `id_token_hint` would be required. It's just recommended, in various places. There is one mention of it being required, but that's in the [document history](https://openid.net/specs/openid-connect-rpinitiated-1_0.html#History) section. But AFAIK the document history section isn't part of the actual specification.

When this PR was sent (and merged), there was a requirement in the specification that `id_token_hint` is required when `post_logout_redirect_uri` is used. You can see that this PR was merged at the end of March.

The latest specification draft from April 2022 does not have this requirement and introduces support for the `client_id` parameter. We changed Keycloak accordingly for the latest draft. This can be expected in Keycloak 19.
